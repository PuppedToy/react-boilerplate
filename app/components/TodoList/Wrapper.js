import styled from 'styled-components';

const Wrapper = styled.div`
  width: 100%;
  height: 400px;
  overflow-y: auto;
`;

export default Wrapper;
