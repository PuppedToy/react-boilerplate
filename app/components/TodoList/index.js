/**
 *
 * TodoList
 *
 */

import React, { memo, useState } from 'react';
import PropTypes from 'prop-types';
// import styled from 'styled-components';

import TodoItem from 'components/TodoItem';

function TodoList({ todos, deleteTodo, toggleTodo }) {
  const [focusedItem, setFocusedItem] = useState(null);

  return (
    <div>
      {todos.map(({ id, title, done }) => (
        <TodoItem
          key={`todo-item-${id}`}
          id={id}
          title={title}
          done={done}
          focused={focusedItem === id}
          focusTodo={() => setFocusedItem(id)}
          deleteTodo={() => deleteTodo(id)}
          toggleTodo={() => toggleTodo(id)}
        />
      ))}
    </div>
  );
}

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  deleteTodo: PropTypes.func,
  toggleTodo: PropTypes.func,
};

export default memo(TodoList);
