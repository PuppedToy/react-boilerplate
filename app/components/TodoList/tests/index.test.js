/**
 *
 * Tests for TodoList
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
// import '@testing-library/jest-dom/extend-expect'; // add some helpful assertions

import TodoList from '../index';

describe('<TodoList />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<TodoList todos={[]} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render empty container on empty todos', () => {
    const { queryAllByRole } = render(<TodoList todos={[]} />);

    expect(queryAllByRole('button').length).toBe(0);
  });

  it('Should render something when there is one todo', () => {
    const todos = [
      {
        id: 1,
        title: 'A todo',
        done: false,
      },
    ];

    const { queryAllByRole } = render(<TodoList todos={todos} />);

    expect(queryAllByRole('button').length).not.toBe(0);
  });

  it('Should focus an item when clicked', () => {
    const todos = [
      {
        id: 1,
        title: 'A todo',
        done: false,
      },
    ];

    const { getByTestId, queryByRole } = render(<TodoList todos={todos} />);

    fireEvent.click(getByTestId('todo-1'));
    expect(queryByRole('textbox')).not.toBe(null);
  });

  it('Should have only the last item focused after clicking another item', () => {
    const todos = [
      {
        id: 1,
        title: 'A todo',
        done: false,
      },
      {
        id: 2,
        title: 'A todo 2',
        done: false,
      },
      {
        id: 3,
        title: 'A todo 3',
        done: false,
      },
    ];

    const { getByTestId, queryByTestId, queryAllByRole } = render(
      <TodoList todos={todos} />,
    );

    fireEvent.click(getByTestId('todo-1'));
    fireEvent.click(getByTestId('todo-3'));

    expect(queryAllByRole('textbox').length).toBe(1);
    expect(queryByTestId('todo-3')).toBe(null);
  });

  it('Should call the delete hook with the appropiate ID when a delete button is clicked', () => {
    const deleteTodoHookMock = jest.fn();

    const todos = [
      {
        id: 1,
        title: 'A todo',
        done: false,
      },
      {
        id: 2,
        title: 'A todo 2',
        done: false,
      },
      {
        id: 3,
        title: 'A todo 3',
        done: false,
      },
    ];

    const { getByTestId } = render(
      <TodoList todos={todos} deleteTodo={deleteTodoHookMock} />,
    );

    fireEvent.click(getByTestId('delete-2'));

    expect(deleteTodoHookMock).toHaveBeenCalledWith(2);
  });

  it('Should call the toggle hook with the appropiate ID when a toggle button is clicked', () => {
    const toggleTodoHookMock = jest.fn();

    const todos = [
      {
        id: 1,
        title: 'A todo',
        done: false,
      },
      {
        id: 2,
        title: 'A todo 2',
        done: false,
      },
      {
        id: 3,
        title: 'A todo 3',
        done: false,
      },
    ];

    const { getByTestId } = render(
      <TodoList todos={todos} toggleTodo={toggleTodoHookMock} />,
    );

    fireEvent.click(getByTestId('toggle-2'));

    expect(toggleTodoHookMock).toHaveBeenCalledWith(2);
  });

  it('Should render and match the snapshot on empty todos', () => {
    const {
      container: { firstChild },
    } = render(<TodoList todos={[]} />);
    expect(firstChild).toMatchSnapshot();
  });

  it('Should render and match the snapshot with one todo', () => {
    const todos = [
      {
        id: 1,
        title: 'A todo',
        done: false,
      },
    ];

    const {
      container: { firstChild },
    } = render(<TodoList todos={todos} />);
    expect(firstChild).toMatchSnapshot();
  });

  it('Should render and match the snapshot with 3 todos', () => {
    const todos = [
      {
        id: 1,
        title: 'A todo',
        done: false,
      },
      {
        id: 2,
        title: 'Another todo',
        done: true,
      },
      {
        id: 3,
        title: 'Last todo',
        done: false,
      },
    ];

    const {
      container: { firstChild },
    } = render(<TodoList todos={todos} />);
    expect(firstChild).toMatchSnapshot();
  });

  it('Should render and match the snapshot with 3 todos after clicking one of them', () => {
    const todos = [
      {
        id: 1,
        title: 'A todo',
        done: false,
      },
      {
        id: 2,
        title: 'Another todo',
        done: true,
      },
      {
        id: 3,
        title: 'Last todo',
        done: false,
      },
    ];

    const {
      container: { firstChild },
      getByTestId,
    } = render(<TodoList todos={todos} />);

    fireEvent.click(getByTestId('todo-3'));
    // @TODO Will this work? Check snapshot when fully implemented!
    expect(firstChild).toMatchSnapshot();
  });
});
