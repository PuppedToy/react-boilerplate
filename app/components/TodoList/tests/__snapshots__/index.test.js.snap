// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> Should render and match the snapshot on empty todos 1`] = `
<div
  class="Wrapper-sc-12k4hfk-0 dilioB"
/>
`;

exports[`<TodoList /> Should render and match the snapshot with 3 todos 1`] = `
<div
  class="Wrapper-sc-12k4hfk-0 dilioB"
>
  <div
    class="Wrapper-ziinx6-0 ecjWnr"
    id="todo-1"
  >
    <div
      class="TodoTitle-xxpseh-0 cHcgZX"
      data-testid="todo-1"
      role="button"
      type="button"
    >
      A todo
    </div>
    <div
      class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
    >
      <div
        alt="check"
        class="Checkbox-eu7a0t-0 hjQtxn"
        data-testid="toggle-1"
        role="button"
        type="button"
      />
      <div
        alt="delete"
        class="DeleteButton-sc-2zb8s8-0 ceLVod"
        data-testid="delete-1"
        role="button"
        type="button"
      />
    </div>
  </div>
  <div
    class="Wrapper-ziinx6-0 ecjWnr"
    id="todo-2"
  >
    <div
      class="TodoTitle-xxpseh-0 cHcgZX"
      data-testid="todo-2"
      role="button"
      type="button"
    >
      Another todo
    </div>
    <div
      class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
    >
      <div
        alt="uncheck"
        class="Checkbox-eu7a0t-0 hjQtxn"
        data-testid="toggle-2"
        role="button"
        type="button"
      />
      <div
        alt="delete"
        class="DeleteButton-sc-2zb8s8-0 ceLVod"
        data-testid="delete-2"
        role="button"
        type="button"
      />
    </div>
  </div>
  <div
    class="Wrapper-ziinx6-0 ecjWnr"
    id="todo-3"
  >
    <div
      class="TodoTitle-xxpseh-0 cHcgZX"
      data-testid="todo-3"
      role="button"
      type="button"
    >
      Last todo
    </div>
    <div
      class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
    >
      <div
        alt="check"
        class="Checkbox-eu7a0t-0 hjQtxn"
        data-testid="toggle-3"
        role="button"
        type="button"
      />
      <div
        alt="delete"
        class="DeleteButton-sc-2zb8s8-0 ceLVod"
        data-testid="delete-3"
        role="button"
        type="button"
      />
    </div>
  </div>
</div>
`;

exports[`<TodoList /> Should render and match the snapshot with 3 todos after clicking one of them 1`] = `
<div
  class="Wrapper-sc-12k4hfk-0 dilioB"
>
  <div
    class="Wrapper-ziinx6-0 ecjWnr"
    id="todo-1"
  >
    <div
      class="TodoTitle-xxpseh-0 cHcgZX"
      data-testid="todo-1"
      role="button"
      type="button"
    >
      A todo
    </div>
    <div
      class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
    >
      <div
        alt="check"
        class="Checkbox-eu7a0t-0 hjQtxn"
        data-testid="toggle-1"
        role="button"
        type="button"
      />
      <div
        alt="delete"
        class="DeleteButton-sc-2zb8s8-0 ceLVod"
        data-testid="delete-1"
        role="button"
        type="button"
      />
    </div>
  </div>
  <div
    class="Wrapper-ziinx6-0 ecjWnr"
    id="todo-2"
  >
    <div
      class="TodoTitle-xxpseh-0 cHcgZX"
      data-testid="todo-2"
      role="button"
      type="button"
    >
      Another todo
    </div>
    <div
      class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
    >
      <div
        alt="uncheck"
        class="Checkbox-eu7a0t-0 hjQtxn"
        data-testid="toggle-2"
        role="button"
        type="button"
      />
      <div
        alt="delete"
        class="DeleteButton-sc-2zb8s8-0 ceLVod"
        data-testid="delete-2"
        role="button"
        type="button"
      />
    </div>
  </div>
  <div
    class="Wrapper-ziinx6-0 ecjWnr"
    id="todo-3"
  >
    <input
      class="Input-tpeqbo-0 VuylU"
      data-testid="editInput-3"
      type="text"
      value="Last todo"
    />
    <div
      class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
    >
      <div
        alt="check"
        class="Checkbox-eu7a0t-0 hjQtxn"
        data-testid="toggle-3"
        role="button"
        type="button"
      />
      <div
        alt="delete"
        class="DeleteButton-sc-2zb8s8-0 ceLVod"
        data-testid="delete-3"
        role="button"
        type="button"
      />
    </div>
  </div>
</div>
`;

exports[`<TodoList /> Should render and match the snapshot with one todo 1`] = `
<div
  class="Wrapper-sc-12k4hfk-0 dilioB"
>
  <div
    class="Wrapper-ziinx6-0 ecjWnr"
    id="todo-1"
  >
    <div
      class="TodoTitle-xxpseh-0 cHcgZX"
      data-testid="todo-1"
      role="button"
      type="button"
    >
      A todo
    </div>
    <div
      class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
    >
      <div
        alt="check"
        class="Checkbox-eu7a0t-0 hjQtxn"
        data-testid="toggle-1"
        role="button"
        type="button"
      />
      <div
        alt="delete"
        class="DeleteButton-sc-2zb8s8-0 ceLVod"
        data-testid="delete-1"
        role="button"
        type="button"
      />
    </div>
  </div>
</div>
`;
